src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,145 @@
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import dataclasses
+
+
+
+from gt4py.next.iterator import ir as itir
+from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.iterator.transforms.trace_shifts import TraceShifts
+from gt4py.next.common import Dimension, DimensionKind
+from gt4py.next.iterator.transforms.global_tmps import SymbolicDomain, SymbolicRange, domain_union
+from gt4py.eve.extended_typing import Dict, List, Tuple, Union
+
+
+# Define a mapping for offset values to dimension names and kinds
+OFFSET_TO_DIMENSION = {
+    itir.SymbolRef('Ioff'): ('IDim', DimensionKind.HORIZONTAL),
+    itir.SymbolRef('Joff'): ('JDim', DimensionKind.HORIZONTAL),
+    itir.SymbolRef('Koff'): ('KDim', DimensionKind.VERTICAL)
+}
+
+
+@dataclasses.dataclass(frozen=True)
+class InferDomain:
+
+    @staticmethod
+    def _infer_dimension_from_offset(offset: itir.OffsetLiteral) -> Dimension:
+        if offset.value in OFFSET_TO_DIMENSION:
+            name, kind = OFFSET_TO_DIMENSION[offset.value]
+            return Dimension(name, kind)
+        else:
+            raise ValueError("offset must be either Ioff, Joff, or Koff")
+
+    @staticmethod
+    def _extract_axis_dims(domain_expr: SymbolicDomain | itir.FunCall) -> List[str]:
+        axis_dims = []
+        if isinstance(domain_expr, SymbolicDomain) and domain_expr.grid_type == 'cartesian_domain':
+            axis_dims.extend(domain_expr.ranges.keys())
+        elif isinstance(domain_expr, itir.FunCall) and domain_expr.fun == im.ref('cartesian_domain'):
+            for named_range in domain_expr.args:
+                if isinstance(named_range, itir.FunCall) and named_range.fun == im.ref('named_range'):
+                    axis_literal = named_range.args[0]
+                    if isinstance(axis_literal, itir.AxisLiteral):
+                        axis_dims.append(axis_literal.value)
+        return axis_dims
+
+    @staticmethod
+    def _get_symbolic_domain(domain: Union[SymbolicDomain, itir.FunCall]) -> SymbolicDomain:
+        if isinstance(domain, SymbolicDomain):
+            return domain
+        if isinstance(domain, itir.FunCall) and domain.fun == im.ref('cartesian_domain'):
+            return SymbolicDomain.from_expr(domain)
+        raise TypeError("domain must either be a FunCall or a SymbolicDomain.")
+
+    @classmethod
+    def _translate_domain(cls, symbolic_domain: SymbolicDomain, shift: Tuple[itir.OffsetLiteral, int],
+                          dims: List[str]) -> SymbolicDomain:
+        new_ranges = {dim: symbolic_domain.ranges[dim] for dim in dims}
+        if shift:
+            off, val = shift
+            current_dim = cls._infer_dimension_from_offset(off)
```

Comment:
You will need the offset provider for connectivities (i.e. unstructured) when you extend this to not only support cartesian dimensions later on. It will be an argument to the final pass that should be forwarded here.
```suggestion
            current_dim = offset_provider[off]
            assert isinstance(current_dim, Dimension)
```
For now it would make sense to
- Add a fixture to your tests that provides you with a cartesian offset provider
  ```
  offset_provider={"IOff": Dimension(...), "JOff": Dimension(...)}
- Pass it as an argument to `infer_as_fieldop` and pass it down the chain until here.